cmd/stdiscosrv/database.go

```diff
@@ -0,0 +1,311 @@
+			// the next.
+			t.Reset(databaseStatisticsInterval)
+
+		case <-s.stop:
+			// We're done.
+			return
+		}
+	}
+}
+
+func (s *levelDBStore) statisticsServe(trigger <-chan struct{}, done chan<- struct{}) {
+	for range trigger {
+		t0 := time.Now()
+		activeCutoff := t0.Add(-24 * time.Hour).UnixNano()
+		active, errors, negative := 0, 0, 0
+
+		iter := s.db.NewIterator(&util.Range{}, nil)
+		for iter.Next() {
+			// Attempt to unmarshal the record, delete it and count the
+			// failure if there's something wrong with it.
+			var rec DatabaseRecord
+			if err := rec.Unmarshal(iter.Value()); err != nil {
+				s.db.Delete(iter.Key(), nil)
+				databaseDeletes.Inc()
+				errors++
+				continue
+			}
+
+			// If there are addresses that have not expired it's an active
+			// record, otherwise it is (or will be on next lookup) a record
+			// representing a lookup failure.
+			if rec.Seen > activeCutoff || len(expire(rec.Addresses, t0.UnixNano())) > 0 {
+				active++
+			} else {
+				negative++
+			}
+		}
+
+		iter.Release()
+
+		databaseKeysActive.Set(float64(active))
+		databaseKeysNegative.Set(float64(negative))
+		databaseKeysError.Set(float64(errors))
+		databaseStatisticsSeconds.Set(time.Since(t0).Seconds())
+
+		// Signal that we are done and can be scheduled again.
+		done <- struct{}{}
+	}
+}
+
+func (s *levelDBStore) Stop() {
+	close(s.stop)
+}
+
+// merge returns the merged result of the two address lists a and b. The
+// result is the union of the two sets, with the newer expiry time chosen
+// for any duplicates.
+func merge(a, b []DatabaseAddress) []DatabaseAddress {
+	// Both lists must be sorted for this to work.
+	sort.Slice(a, func(i, j int) bool {
+		return a[i].Address < a[j].Address
+	})
+	sort.Slice(b, func(i, j int) bool {
+		return b[i].Address < b[j].Address
+	})
+
+	aIdx := 0
+	bIdx := 0
+	res := make([]DatabaseAddress, 0, len(a)+len(b))
+
+loop:
+	for {
+		switch {
+		case aIdx == len(a) && bIdx == len(b):
+			// both lists are exhausted, we are done
+			break loop
+
+		case aIdx == len(a):
+			// a is exhausted, pick from b and continue
+			res = append(res, b[bIdx])
+			bIdx++
+			continue
+
+		case bIdx == len(b):
+			// b is exhausted, pick from a and continue
+			res = append(res, a[aIdx])
+			aIdx++
+			continue
+		}
+
+		// We have values left on both sides.
+		aVal := a[aIdx]
+		bVal := b[bIdx]
+
+		switch {
+		case aVal.Address == bVal.Address:
+			// update for same address, pick newer
+			if aVal.Expires > bVal.Expires {
+				res = append(res, aVal)
+			} else {
+				res = append(res, bVal)
+			}
+			aIdx++
+			bIdx++
+
+		case aVal.Address < bVal.Address:
+			// a is smallest, pick it and continue
+			res = append(res, aVal)
+			aIdx++
+
+		default:
+			// b is smallest, pick it and continue
+			res = append(res, bVal)
+			bIdx++
+		}
+	}
+	return res
+}
+
+// expire returns the list of addresses after removing expired entries
+func expire(addrs []DatabaseAddress, now int64) []DatabaseAddress {
```

Comment:
Again, in place delete, like in other places would be better, but this is nitpicking. Gotta shave off those uS for save costs on infra.